src/libcore/intrinsics.rs

```diff
@@ -1031,24 +1097,65 @@ extern "rust-intrinsic" {
     ///     dst
     /// }
     /// ```
-    ///
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn copy<T>(src: *const T, dst: *mut T, count: usize);
 
-    /// Invokes memset on the specified pointer, setting `count * size_of::<T>()`
-    /// bytes of memory starting at `dst` to `val`.
+    /// Sets `count * size_of::<T>()` bytes of memory starting at `dst` to
+    /// `val`.
+    ///
+    /// `write_bytes` is similar to C's [`memset`], but sets `count *
+    /// size_of::<T>()` bytes to `val`.
+    ///
+    /// [`memset`]: https://www.gnu.org/software/libc/manual/html_node/Copying-Strings-and-Arrays.html#index-memset
+    ///
+    /// # Safety
+    ///
+    /// Behavior is undefined if any of the following conditions are violated:
+    ///
+    /// * `dst` must be [valid] for writes of `count * size_of::<T>()` bytes.
+    ///
+    /// * `dst` must be properly aligned.
+    ///
+    /// Additionally, the caller must ensure that writing `count *
+    /// size_of::<T>()` bytes to the given region of memory results in a valid
+    /// value of `T`. Creating an invalid value of `T` can result in undefined
+    /// behavior.
```

Comment:
I don't know how to put this better than you did, but the first sentence says that this "must" be ensured, while the second sentence says that not ensuring it "can" be UB. 

_I think_ we are pretty sure that, if a region of memory that does not contain a valid value of `T` is "used" as a `T`, that is UB. What we are not so sure about is whether this value existing is already UB (e.g. "copy invalid byte pattern to it, then copy it somewhere else, then set a valid byte patter, then use the value with the valid byte pattern" might work). 

Maybe a warning here would be better, something along the lines of:

> Additionally, the caller should be careful about the value contained in the region of memory after the `memset` and/or how this region of memory is subsequently used since using a region of memory typed as a `T` that contains an invalid value of `T` is undefined behavior.

Reply:
I picked some hybrid of what you proposed:

> Additionally, the caller should ensure that writing `count * size_of::<T>()` bytes to the given region of memory results in a valid value of `T`. Using a region of memory typed as a `T` that contains an invalid value of `T` is undefined behavior.